File: src/tqi/gtfs/filter.py
"""Filter GTFS feed to Chilliwack routes and a target service day."""

from datetime import date, datetime, timedelta

import pandas as pd

from tqi.config import CHILLIWACK_ROUTES
from tqi.gtfs.parse import GTFSFeed

WEEKDAYS = ["monday", "tuesday", "wednesday", "thursday", "friday"]


def _find_best_weekday(
    calendar: pd.DataFrame | None,
    calendar_dates: pd.DataFrame | None,
    trips: pd.DataFrame,
) -> tuple[str, date]:
    """Find the weekday with the most active trips.

    Handles both calendar.txt and calendar_dates.txt-only feeds.
    Returns (day_name, reference_date).
    """
    best_day = "wednesday"
    best_date = date.today()
    best_trips = 0

    # Collect candidate dates from calendar_dates.txt
    candidate_dates: list[date] = []

    if calendar_dates is not None and not calendar_dates.empty:
        for d_str in calendar_dates["date"].unique():
            try:
                d = datetime.strptime(str(d_str), "%Y%m%d").date()
                if d.weekday() < 5:  # weekdays only
                    candidate_dates.append(d)
            except ValueError:
                continue

    # Also try dates from calendar.txt range
    if calendar is not None and not calendar.empty and "start_date" in calendar.columns:
        start = datetime.strptime(calendar["start_date"].min(), "%Y%m%d").date()
        end = datetime.strptime(calendar["end_date"].max(), "%Y%m%d").date()
        d = start
        while d <= end and len(candidate_dates) < 100:
            if d.weekday() < 5:
                candidate_dates.append(d)
            d += timedelta(days=1)

    if not candidate_dates:
        # Last resort: try the next 7 days from today
        for i in range(7):
            d = date.today() + timedelta(days=i)
            if d.weekday() < 5:
                candidate_dates.append(d)

    # For each candidate date, count how many trips are active
    for d in candidate_dates:
        d_str = d.strftime("%Y%m%d")
        day_name = WEEKDAYS[d.weekday()]

        active_services: set[str] = set()

        # calendar.txt
        if calendar is not None and not calendar.empty:
            mask = calendar[day_name] == "1"
            if "start_date" in calendar.columns:
                mask = mask & (calendar["start_date"] <= d_str) & (calendar["end_date"] >= d_str)
            active_services = set(calendar.loc[mask, "service_id"])

        # calendar_dates.txt exceptions
        if calendar_dates is not None and not calendar_dates.empty:
            for _, row in calendar_dates[calendar_dates["date"] == d_str].iterrows():
                if row["exception_type"] == "1":
                    active_services.add(row["service_id"])
                elif row["exception_type"] == "2":
                    active_services.discard(row["service_id"])

        if active_services:
            n_trips = len(trips[trips["service_id"].isin(active_services)])
            if n_trips > best_trips:
                best_trips = n_trips
                best_day = day_name
                best_date = d

    return best_day, best_date


def filter_to_chilliwack(
    feed: GTFSFeed,
    target_day: str | None = None,
    routes: list[str] = CHILLIWACK_ROUTES,
) -> GTFSFeed:
    """Filter feed to Chilliwack routes on the best weekday.

    If target_day is None, automatically picks the weekday with the most trips.
    """
    if target_day is None:
        day_name, ref_date = _find_best_weekday(
            feed.calendar, feed.calendar_dates, feed.trips
        )
        print(f"Auto-selected best weekday: {day_name} ({ref_date})")
    else:
        day_name = target_day
        # Find a valid date for the requested day
        _, ref_date = _find_best_weekday(feed.calendar, feed.calendar_dates, feed.trips)
        # Override to find a date matching target_day
        day_index = WEEKDAYS.index(target_day) if target_day in WEEKDAYS else 2
        if ref_date.weekday() != day_index:
            # Adjust to nearest matching weekday
            diff = (day_index - ref_date.weekday()) % 7
            ref_date = ref_date + timedelta(days=diff)
        print(f"Using reference date: {ref_date} ({day_name})")

    # Resolve active services for that date
    active_services: set[str] = set()
    ref_str = ref_date.strftime("%Y%m%d")

    if feed.calendar is not None and not feed.calendar.empty:
        mask = feed.calendar[day_name] == "1"
        if "start_date" in feed.calendar.columns:
            mask = mask & (feed.calendar["start_date"] <= ref_str) & (feed.calendar["end_date"] >= ref_str)
        active_services = set(feed.calendar.loc[mask, "service_id"])

    if feed.calendar_dates is not None and not feed.calendar_dates.empty:
        for _, row in feed.calendar_dates[feed.calendar_dates["date"] == ref_str].iterrows():
            if row["exception_type"] == "1":
                active_services.add(row["service_id"])
            elif row["exception_type"] == "2":
                active_services.discard(row["service_id"])

    if not active_services:
        raise ValueError(f"No active services found for {day_name} on {ref_date}")
    print(f"Active services: {len(active_services)} ({', '.join(sorted(active_services))})")

    # Filter routes by short_name
    route_mask = feed.routes["route_short_name"].isin(routes)
    filtered_routes = feed.routes[route_mask]
    route_ids = set(filtered_routes["route_id"])
    print(f"Chilliwack routes matched: {len(route_ids)}")

    # Filter trips
    trip_mask = feed.trips["route_id"].isin(route_ids) & feed.trips["service_id"].isin(
        active_services
    )
    filtered_trips = feed.trips[trip_mask]
    trip_ids = set(filtered_trips["trip_id"])
    print(f"Active trips: {len(trip_ids)}")

    # Filter stop_times
    filtered_stop_times = feed.stop_times[feed.stop_times["trip_id"].isin(trip_ids)]

    # Filter stops to only those referenced
    used_stop_ids = set(filtered_stop_times["stop_id"])
    filtered_stops = feed.stops[feed.stops["stop_id"].isin(used_stop_ids)]
    print(f"Stops in use: {len(filtered_stops)}")

    return GTFSFeed(
        stops=filtered_stops.reset_index(drop=True),
        stop_times=filtered_stop_times.reset_index(drop=True),
        trips=filtered_trips.reset_index(drop=True),
        routes=filtered_routes.reset_index(drop=True),
        calendar=feed.calendar,
        calendar_dates=feed.calendar_dates,
        shapes=feed.shapes,
    )

File: go/internal/gtfs/filter.go
package gtfs

import (
	"strings"
	"time"
)

var weekdays = []string{"monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"}

// FindBestWeekday scans the calendar for the weekday with the most trips.
// Returns the day name (lowercase) and a reference date for that day.
func FindBestWeekday(feed *Feed) (string, time.Time) {
	// Count trips per service
	tripCount := make(map[string]int)
	for _, t := range feed.Trips {
		tripCount[t.ServiceID]++
	}

	// For each day, sum trips from all services active that day
	bestDay := ""
	bestCount := 0
	for _, day := range weekdays {
		count := 0
		for _, cal := range feed.Calendar {
			if calendarDayActive(cal, day) {
				count += tripCount[cal.ServiceID]
			}
		}
		if count > bestCount {
			bestCount = count
			bestDay = day
		}
	}

	if bestDay == "" {
		bestDay = "monday"
	}

	// Find a reference date for this day. Use the first calendar entry's
	// start date and advance to the target weekday.
	refDate := findReferenceDate(feed, bestDay)

	return bestDay, refDate
}

// findReferenceDate returns a date within the calendar range that falls on
// the given weekday.
func findReferenceDate(feed *Feed, day string) time.Time {
	targetWd := dayToWeekday(day)

	if len(feed.Calendar) > 0 {
		start, err := time.Parse("20060102", feed.Calendar[0].StartDate)
		if err == nil {
			// Advance to the target weekday
			for i := 0; i < 7; i++ {
				d := start.AddDate(0, 0, i)
				if d.Weekday() == targetWd {
					return d
				}
			}
		}
	}
	// Fallback: use next occurrence from today
	now := time.Now()
	for i := 0; i < 7; i++ {
		d := now.AddDate(0, 0, i)
		if d.Weekday() == targetWd {
			return d
		}
	}
	return now
}

func dayToWeekday(day string) time.Weekday {
	switch strings.ToLower(day) {
	case "monday":
		return time.Monday
	case "tuesday":
		return time.Tuesday
	case "wednesday":
		return time.Wednesday
	case "thursday":
		return time.Thursday
	case "friday":
		return time.Friday
	case "saturday":
		return time.Saturday
	case "sunday":
		return time.Sunday
	}
	return time.Monday
}

// FilterFeed filters a feed to only the given routes and the target day's services.
func FilterFeed(feed *Feed, routes []string, targetDay string) (*Feed, error) {
	// Resolve active services for the target day
	activeServices := make(map[string]bool)
	for _, cal := range feed.Calendar {
		if calendarDayActive(cal, targetDay) {
			activeServices[cal.ServiceID] = true
		}
	}

	// TODO: handle calendar_dates exceptions if needed in the future

	// Filter routes by short name
	wantRoutes := make(map[string]bool)
	for _, r := range routes {
		wantRoutes[r] = true
	}

	var filteredRoutes []Route
	routeIDs := make(map[string]bool)
	for _, r := range feed.Routes {
		if wantRoutes[r.RouteShortName] {
			filteredRoutes = append(filteredRoutes, r)
			routeIDs[r.RouteID] = true
		}
	}

	// Filter trips by route + service
	var filteredTrips []Trip
	tripIDs := make(map[string]bool)
	for _, t := range feed.Trips {
		if routeIDs[t.RouteID] && activeServices[t.ServiceID] {
			filteredTrips = append(filteredTrips, t)
			tripIDs[t.TripID] = true
		}
	}

	// Filter stop_times by trip
	var filteredStopTimes []StopTime
	usedStopIDs := make(map[string]bool)
	for _, st := range feed.StopTimes {
		if tripIDs[st.TripID] {
			filteredStopTimes = append(filteredStopTimes, st)
			usedStopIDs[st.StopID] = true
		}
	}

	// Filter stops to only those used
	var filteredStops []Stop
	for _, s := range feed.Stops {
		if usedStopIDs[s.StopID] {
			filteredStops = append(filteredStops, s)
		}
	}

	return &Feed{
		Stops:         filteredStops,
		StopTimes:     filteredStopTimes,
		Trips:         filteredTrips,
		Routes:        filteredRoutes,
		Calendar:      feed.Calendar,
		CalendarDates: feed.CalendarDates,
		Shapes:        feed.Shapes,
	}, nil
}

// calendarDayActive checks if a calendar entry is active on the given day.
func calendarDayActive(cal Calendar, day string) bool {
	switch strings.ToLower(day) {
	case "monday":
		return cal.Monday
	case "tuesday":
		return cal.Tuesday
	case "wednesday":
		return cal.Wednesday
	case "thursday":
		return cal.Thursday
	case "friday":
		return cal.Friday
	case "saturday":
		return cal.Saturday
	case "sunday":
		return cal.Sunday
	}
	return false
}
